visibility enables possible name clashes

one example:
```
~/playground/fz 1$ find . -type f -exec echo {} \; -exec cat {} \;
./extern.fz
module extern is
  private feat String is
    "Hello World"
  module call_feat String is
    feat
./test.fz
ex : extern is

  feat => 2

  say feat
  say call_feat
```
the crashes:
```
~/playground/fz 130$ fz -c -sourceDirs=/home/sam/playground/fz ~/playground/fz/test.fz 
ex.c:18251:26: error: cast to 'fzT__RString *' (aka 'struct fzT__RString *') from smaller integer type 'fzT_1i32' (aka 'int') [-Werror,-Wint-to-pointer-cast]
    fzCur.fzF_1_result = (fzT__RString*)fzM_0;
                         ^~~~~~~~~~~~~~~~~~~~
1 error generated.

error 1: C backend: C compiler failed
C compiler call 'clang -Wall -Werror -Wno-gnu-empty-struct -Wno-unused-variable -Wno-unused-label -Wno-unused-function -Wno-infinite-recursion -Wno-unused-but-set-variable -O3 -lm -lpthread -o ex ex.c' failed with exit code '1'

one error.
~/playground/fz 1$ fz -sourceDirs=/home/sam/playground/fz ~/playground/fz/test.fz 
2

error 1: *** java.lang.Error: check-condition failed: dev.flang.be.interpreter.ValueWithClazz:clazz:90
Call stack:
(((io.#type io).out.#type io.out).default_print_handler).println: $FUZION/lib/io/print_effect.fz:55:14:
    print (s.as_string + (codepoint 10))
-------------^
io.out.println: $FUZION/lib/io/print_effect.fz:36:7:
    p.println s
------^
say: $FUZION/lib/say.fz:31:29:
public say(s Any) => io.out.println s
----------------------------^
ex: /home/sam/playground/fz/test.fz:6:3:
  say call_feat
--^^^



error 2: java.lang.Error: check-condition failed: dev.flang.be.interpreter.ValueWithClazz:clazz:90
	at dev.flang.util.ANY.check(ANY.java:383)
	at dev.flang.be.interpreter.ValueWithClazz.clazz(ValueWithClazz.java:90)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:327)
	at dev.flang.be.interpreter.Interpreter.executeArgs(Interpreter.java:615)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:297)
	at dev.flang.be.interpreter.Interpreter.executeArgs(Interpreter.java:620)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:297)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:343)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:388)
	at dev.flang.be.interpreter.Interpreter.callOnInstance(Interpreter.java:905)
	at dev.flang.be.interpreter.Interpreter.lambda$callable$6(Interpreter.java:781)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:332)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:388)
	at dev.flang.be.interpreter.Interpreter.executeArgs(Interpreter.java:615)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:297)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:507)
	at dev.flang.be.interpreter.Interpreter.executeArgs(Interpreter.java:615)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:297)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:343)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:388)
	at dev.flang.be.interpreter.Interpreter.callOnInstance(Interpreter.java:905)
	at dev.flang.be.interpreter.Interpreter.lambda$callable$6(Interpreter.java:781)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:332)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:343)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:388)
	at dev.flang.be.interpreter.Interpreter.callOnInstance(Interpreter.java:905)
	at dev.flang.be.interpreter.Interpreter.lambda$callable$6(Interpreter.java:781)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:332)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:388)
	at dev.flang.be.interpreter.Interpreter.callOnInstance(Interpreter.java:905)
	at dev.flang.be.interpreter.Interpreter.lambda$callable$6(Interpreter.java:781)
	at dev.flang.be.interpreter.Interpreter.run(Interpreter.java:228)
	at dev.flang.tools.Fuzion$Backend$1.process(Fuzion.java:106)
	at dev.flang.tools.Fuzion$Backend.processFrontEnd(Fuzion.java:406)
	at dev.flang.tools.Fuzion.lambda$parseArgsForBackend$3(Fuzion.java:870)
	at dev.flang.tools.Tool.lambda$run$0(Tool.java:156)
	at dev.flang.util.Errors.runAndExit(Errors.java:748)
	at dev.flang.tools.Tool.run(Tool.java:156)
	at dev.flang.tools.Fuzion.main(Fuzion.java:519)


*** fatal errors encountered, stopping.
```

--- a/src/dev/flang/be/c/CNames.java
+++ b/src/dev/flang/be/c/CNames.java
@@ -27,6 +27,7 @@ Fuzion language implementation.  If not, see <https://www.gnu.org/licenses/>.
 package dev.flang.be.c;
 
 import java.util.ArrayList;
+import java.util.TreeSet;
 
 import dev.flang.fuir.FUIR;
 
@@ -252,6 +253,12 @@ public class CNames extends ANY
     private final ArrayList<String> _cache = new ArrayList<>();
 
 
+    /*
+     * Set of used names
+     */
+    private final TreeSet<String> _cacheSet = new TreeSet<String>();
+
+
     /**
      * prefix to be used for given class of names.
      */
@@ -335,6 +342,17 @@ public class CNames extends ANY
               if (CHECKS) check
                 (res.length() == MAX_C99_IDENTIFIER_LENGTH);
             }
+          // see reg_issue1840 why the following is needed
+          var i = 0;
+          if (_cacheSet.contains(res))
+            {
+              while(_cacheSet.contains(res + i))
+                {
+                  i++;
+                }
+              res = res + i;
+            }
+          _cacheSet.add(res);
           _cache.set(num, res);
         }
 

--- a/src/dev/flang/be/c/CNames.java
+++ b/src/dev/flang/be/c/CNames.java
@@ -256,7 +256,7 @@ public class CNames extends ANY
     /*
      * Set of used names
      */
-    private final TreeSet<String> _cacheSet = new TreeSet<String>();
+    private final TreeSet<String> _usedNames = new TreeSet<String>();
 
 
     /**
@@ -327,11 +327,15 @@ public class CNames extends ANY
           var p = _prefix;
           var sb = new StringBuilder(p);
           clazzMangledName(cl, sb);
-          // NYI: there might be name conflicts due to different generic instances, so
-          // we need to add the clazz id or the actual generics if this is the case:
-          //
-          //   sb.append("_").append(clazzId2num(cl)).append("_");
           res = sb.toString();
+
+          // see reg_issue1840 why the following is needed
+          // name clashes due to equal named features that are not visible but inherited
+          if (_usedNames.contains(res))
+          {
+            res = res + "_" + clazzId2num(cl);
+          }
+          _usedNames.add(res);
 
           if (res.length() > MAX_C99_IDENTIFIER_LENGTH)
             {
@@ -342,17 +346,6 @@ public class CNames extends ANY
               if (CHECKS) check
                 (res.length() == MAX_C99_IDENTIFIER_LENGTH);
             }
-          // see reg_issue1840 why the following is needed
-          var i = 0;
-          if (_cacheSet.contains(res))
-            {
-              while(_cacheSet.contains(res + i))
-                {
-                  i++;
-                }
-              res = res + i;
-            }
-          _cacheSet.add(res);
           _cache.set(num, res);
         }
 

--- a/tests/reg_issue1840/skip_c
+++ b/tests/reg_issue1840/skip_c
@@ -1,1 +1,0 @@
-NYI: visibility enables possible name clashes

